kitsune/wiki/utils.py

```diff
@@ -182,26 +182,53 @@ def get_visible_document_or_404(
     except Document.DoesNotExist:
         pass
 
-    if (
-        not look_for_translation_via_parent
-        or not (locale := kwargs.get("locale"))
-        or (locale == settings.WIKI_DEFAULT_LANGUAGE)
-    ):
-        # We either don't want to try to find the translation via its parent, or it doesn't
-        # make sense, because we're not making a locale-specific request or the locale we're
-        # requesting is already the default locale.
+    if not (locale := kwargs.get("locale")):
         raise Http404
 
-    # We couldn't find a visible translation in the requested non-default locale, so let's
-    # see if we can find a visible translation via its parent.
+    slug = kwargs.get("slug")
+
+    if slug and look_for_translation_via_parent:
+        # Find documents with this slug in other locales
+        other_docs_qs = (
+            Document.objects.filter(slug=slug).exclude(locale=locale).select_related("parent")
+        )
+
+        if locale == settings.WIKI_DEFAULT_LANGUAGE:
+            # We're in default locale, look for translations in other locales
+            for doc in other_docs_qs:
+                parent_doc = doc.parent or doc
+
+                # If the parent is in the default locale, return it directly
+                if parent_doc.locale == settings.WIKI_DEFAULT_LANGUAGE:
+                    return parent_doc
+
+                # Otherwise check if it has a translation in the requested locale
+                translation = parent_doc.translated_to(locale, visible_for_user=user)
+                if translation:
+                    return translation
+        else:
+            # Looking for a non-default locale document
+            for doc in other_docs_qs:
+                # Only consider documents that are translations
+                if not doc.parent:
+                    continue
+
+                # Check if this parent has a translation in our requested locale
+                translation = doc.parent.translated_to(locale, visible_for_user=user)
+                if translation:
+                    return translation
+
+    if not look_for_translation_via_parent or locale == settings.WIKI_DEFAULT_LANGUAGE:
```

Comment:
Is there a reason this check is so late in the code after all the expensive operations above? This should be combined with the earlier 404.

Reply:
Yes - the second check after the loop prevents the final fallback when we're in the default language. Checking it earlier breaks the case where a user requests a document in one locale but is using a slug from another. So - check to see if we should bother finding a translation.